### PR TITLE
Microbenchmarking

### LINKED ISSUE #78: Microbenchmarking
* one level microbenchmark on spaces: test index lookup speed and so on
* two level microbenchmark on shard: test network throughput utilization
* three level microbenchmark on shard with 100 matching objects on the first level, many ones on the second level and 1 on the third level

### PR DESCRIPTION
**Pull request by Totktonada**
Closes #78.

### CODE REVIEW

File: test/common/lua/multirunner.lua [hunk 1]
```diff
@@ -1,15 +1,61 @@
...
+local net_box = require('net.box')
+local utils = require('graphql.utils')
 
 local function instance_uri(instance_id)
     local socket_dir = require('fio').cwd()
     local uri = ('%s/shard%s.sock'):format(socket_dir, instance_id)
     return uri
 end
 
+local SERVERS_DEFAULT = {'shard1', 'shard2', 'shard3', 'shard4'}
```
Comment (SudoBobo) [inline test/common/lua/multirunner.lua]: SERVERS_DEFAULT and SHARD_CONF_DEFAULT.servers are a little bit confusing

File: test/common/lua/multirunner.lua [hunk 2]
```diff
@@ -40,75 +86,83 @@ local function for_each_server(shard, func)
...
+    local callback = opts.callback
+
+    assert(test_run ~= nil)
+    assert(init_function ~= nil)
+    assert(cleanup_function ~= nil)
+    assert(callback ~= nil)
+
+    if conf_type == 'space' then
+        init_function()
+        callback("Local (box)", nil)
```
Comment (SudoBobo) [inline test/common/lua/multirunner.lua]: why not `callback(conf_name, nil)`?

File: test/common/lua/multirunner.lua [hunk 3]
```diff
@@ -40,75 +86,83 @@ local function for_each_server(shard, func)
...
+        local cleanup_script = string.dump(cleanup_function)
+
+        local shard = init_shard(test_run, servers, shard_conf)
+
+        for_each_server(shard, function(uri)
+            local c = net_box.connect(uri)
+            c:eval(init_script)
+        end)
+
+        callback(conf_name, shard)
```
Comment (SudoBobo) [inline test/common/lua/multirunner.lua]: Are you sure that this function should be called `callback`? It is not called after any actions  so it should not be called like that. Maybe `command` as it is something we want from our shards to run?

File: test/bench/nesting-1.test.lua [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    for i = 2, ITERATIONS do
+        local result = yaml.encode(gql_query_1:execute(variables_1))
+        checksum:update(result .. i)
+        fiber.yield()
+    end
+
+    local end_time = clock.monotonic64()
+
+    print('Iterations: ' .. tostring(ITERATIONS))
+    print('Checksum: ' .. tostring(checksum:result()))
```
Comment (SudoBobo) [inline test/bench/nesting-1.test.lua]: What for do we need `Checksum`?

File: test/bench/nesting-1.test.lua [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    local rps = ITERATIONS / duration
+
+    return {
+        duration = duration,
+        latency = latency,
+        rps = rps,
+    }
+end
+
+local function octal(...)
```
Comment (SudoBobo) [inline test/bench/nesting-1.test.lua]: Move to test/utils?

File: test/bench/nesting-1.test.lua (hunk 4 above)
Comment (Totktonada) [reply test/bench/nesting-1.test.lua]: Result correctness check. W/o a checksum verification we can measure something other then we expected.

File: test/bench/nesting-1.test.lua (hunk 5 above)
Comment (Totktonada) [reply test/bench/nesting-1.test.lua]: Will use `tonumber('644', 8)`.

File: test/common/lua/multirunner.lua (hunk 3 above)
Comment (Totktonada) [reply test/common/lua/multirunner.lua]: @Khatskevich is the original author.

File: test/common/lua/multirunner.lua (hunk 2 above)
Comment (Totktonada) [reply test/common/lua/multirunner.lua]: You are right, fixed.

File: test/common/lua/multirunner.lua (hunk 3 above)
Comment (Khatskevich) [reply test/common/lua/multirunner.lua]: I do not insist. It seems that callback is not the best option, however, I suppose that this small detail does not deserve as much attention.

Comment (Totktonada) [reply test/common/lua/multirunner.lua]: Will change to `workload`.

File: test/common/lua/multirunner.lua (hunk 1 above)
Comment (Totktonada) [reply test/common/lua/multirunner.lua]: Added comments.

File: test/bench/nesting-1.test.lua (hunk 4 above)
Comment (Totktonada) [reply test/bench/nesting-1.test.lua]: Such approach is used in compiler benchmarks to make sure the compiler does not omit executing of 'dead' (unused) code of benchmark. So we need to print entire output or checksum. Print possibly large output can change timings significantly, so we stick to checksum approach.

**Comment by Totktonada** [comment]
Split workload from benchmarking loop code.

**Comment by Totktonada** [comment]
Added planned set of microbenchmarks.

**Comment by Totktonada** [comment]
Increased time limit for requests in bench/*.test.lua.

**Review by SudoBobo — APPROVED** [review]

**Pull request merged**